### PR TITLE
Foms

### PR DESCRIPTION
**Pull request by xrasskin**
Created a new combined version of the nema7 and jaszczak foms python wrapper and called it `foms.py`.

### CODE REVIEW

**Comment by jacg** [comment]
The original `jaszczak_foms.py` and `nema7foms.py` need to be removed, too.

**Comment by jacg** [comment]
As discussed, we also want to add an option for plotting 6 separate graphs (one for each sphere) vs. all 6 spheres on a single graph.

On the graphs, we decided to

+ use the subset numbers as labels on the x-axis
+ draw vertical lines for visual accentuation of subsets

**Comment by jacg** [comment]
The point of the `TODO` about adding ability to read foms from generated fom files, was to avoid recalculating the foms if you want to show the plot. As it stands, if I calculate the foms (and the file is written), and then decide I want to look at the graphs, and add the `--show-plot=True` option, it will calculate all the figures all over again, wasting a lot of time.

So:

+  If `foms` exists in the directory:
    -  just read the data from the file, rather than calculating them again
    -  print a message at the end, stating that the figures were taken from the cache file, and that the user should delete the file if the figures are to be calculated again (the images might have been regenerated, since the last time)
+ If `foms` is absent:
   - calculate the foms
   - print them out as they are being calculated (as happens already)
   - write the foms file

File: src/foms.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        commandStart = 'target/release/foms -- nema7 '
+        write_command(filename, images, commandStart, nema7_sphere_diameters,
+                      cli_args)
+        plot_from_fom(filename, nema7_sphere_diameters, cli_args)
+
+    else:
+        print("You must specify a known phantom type: (jaszczak/nema7)")
+
+if __name__=='__main__':
+   main()
```
Comment (jacg) [inline src/foms.py:144]: I'm no fan of these pointless guards which add no useful functionality of the code.

I'm aware that > 90% of Python programmers stick these in by default, and some (many?, most?) IDEs do too, but that's not a good enough reason for me.

Do you know what this does? What it's purpose is?

These are useful when you want the same source file to be usable both as a module (something you `import`) and as program (something you 'execute'). This construct allows you to distinguish between the two situations from within the code, and do things slightly differently. In brief

+ The program should probably do some work, produce some output, etc.
+ The module should most probably *not* produce any output or do any work, besides defining the utilities that the `import`er wants to use.

More official details here: https://docs.python.org/3/library/__main__.html

This particular source file is only intended to be used as a program, and defines nothing that might be useful to import elsewhere, so let's not pretend that it's otherwise by including this guard.

I'm also no fan of pointless `main` functions. Python programs are executed from top to bottom, so you don't need a `main` to tell the system what the entry point is (unlike C, Rust, etc.).

I'm less negative about pointless `main`s than I am about pointless `__main__` guards, as there might be *some* value in grouping or hiding stuff in functions. You don't pollute the global scope with `main`'s local variables, for one.

In short, you can keep `main()`, if you like, but please remove the guard.

File: src/foms.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        plt.show()
+
+def main():
+    cli_args = docopt(__doc__)
+    directory = cli_args['<DIR>']
+    print(f'Calculating FOMs for images in {directory}')
+    filename = f'{directory}/foms'
+    images = sorted(Path(directory).glob('*.raw'))
+
+    if bool(cli_args['<jaszczak>']):
```
Comment (jacg) [inline src/foms.py:128]: The `bool` is superfluous.

`cli_args['<jaszczak>']` will evaluate to one of

+ `'jaszczak'`
+ `None`

Non-empty strings are true in boolean contexts, and `None` is false in boolean contexts. This is by design, so idiomatic Python use takes full advantage of these features.

For example, don't write 

```python
if len(my_sequence) != 0:
```
nor
```python
if len(my_sequence):
```
but
```python
if my_sequence:
```

All 3 have the same meaning as far as the flow of control is concerned, but the last one is idiomatic Python.

If you're not sure about Python's boolean punning, let's talk about it.

File: src/foms.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        commandStart = 'target/release/foms -- jaszczak '
+        write_command(filename, images, commandStart, jaszczak_sphere_diameters,
+                      cli_args)
+        plot_from_fom(filename, jaszczak_sphere_diameters, cli_args)
+
+    elif bool(cli_args['<nema7>']):
+        commandStart = 'target/release/foms -- nema7 '
```
Comment (jacg) [inline src/foms.py:135]: It turns out that the trailing space in `commandStart` is *crucial* for the correct functioning of the code.

This is bad. Let's try to remove this 'feature'.

Comment (jacg) [reply src/foms.py:135]: ... and I suggest you do it like this:
```diff
- foms = get_foms(command + str(image_file_name))
+ foms = get_foms(f'{command} {image_file_name}')
```
in the definition of `write_command`.

File: src/foms.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        write_command(filename, images, commandStart, nema7_sphere_diameters,
+                      cli_args)
+        plot_from_fom(filename, nema7_sphere_diameters, cli_args)
```
[the remark this answers is not in the document]
Comment (jacg) [reply src/foms.py:138]: ... or, just for fun

```python
    phantom = (
        cli_args['<jaszczak>'] or
        cli_args['<nema7>'   ] or
        sys.exit("You must specify a known phantom type: (jaszczak/nema7)")
    )
```
:wink:

[the remark this answers is not in the document]
Comment (jacg) [reply src/foms.py:138]: Actually, 
```python
phantom = cli_args['<jaszczak>'] or cli_args['<nema7>']
```
is enough.

I'll let you figure out why.

**Comment by jacg** [comment]
I've rebased this branch on top of the latest master.
